### PR TITLE
feat: add tool calling support to conversation api

### PR DESCRIPTION
**Pull request by sicoyle**
# Description

For 1.16 we want tool calling support added to the conversation api. Corresponding backend runtime changes are here https://github.com/dapr/dapr/pull/8880. I made the decision to keep the translation from our proto inputs directly to the langchain go sdk structs, so I've updated all my structs here with the new conversation api version and corresponding langchain pkg types to help reduce translating all the fields. This also enables us to use the langchain go structs and then for the one component that is different (deep seek), I can easily translate to just that outlier sdk type from the langchain go struct fields.

I've added conformance tests and unit test updates for echo component so far. Am working on rest to see if there are edge cases I need to acct for.

I also split the files up according to the old types vs the new api types to help clean things up for us.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #_[issue number]_

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: conversation/langchaingokit/message.go [hunk 1]
```diff
@@ -37,13 +37,34 @@ func GetMessageFromRequest(r *conversation.ConversationRequest) []llms.MessageCo
...
+		opts = make([]llms.CallOption, 0)
+	}
+
+	if r.Temperature > 0 {
+		opts = append(opts, llms.WithTemperature(r.Temperature))
+	}
+
+	if r.Tools != nil {
+		// TODO: go back and maybe make this a pointer to slice of tools instead?
+		tools := make([]llms.Tool, len(r.Tools))
```
Comment (JoshVanL) [inline conversation/langchaingokit/message.go]: Why can we not match the type?

File: conversation/conversev1alpha2.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	Id string `json:"id"`
+	ToolCallFunction
+}
+
+type ToolCallFunction struct {
+	Name      string  `json:"name"`
+	Arguments *string `json:"arguments"` // can be empty
+}
+
+type ConversationRequestV1Alpha2 struct {
```
Comment (JoshVanL) [inline conversation/conversev1alpha2.go]: Rather than needing a v2, can we not transform a v1 public API message  into the v2 internal API?

**Review by JoshVanL — CHANGES_REQUESTED** [review]

File: conversation/conversev1alpha2.go (hunk 2 above)
Comment (sicoyle) [reply conversation/conversev1alpha2.go]: i made my contrib v2 structs be the v1 structs transforming all protos into the new structs i've got now in contrib with my latest commit. Lmk if you were wanting something diff but that cleaned things up nicely.

File: tests/conformance/README.md [hunk 3]
```diff
@@ -90,14 +90,24 @@
 To run all conformance tests
 
 ```bash
- dlv test --build-flags '-v -tags=conftests' ./tests/conformance
+dlv test --build-flags '-v -tags=conftests' ./tests/conformance
```
Comment (daixiang0) [inline tests/conformance/README.md:94]: Add go test case as well.

File: tests/certification/go.mod [hunk 4]
```diff
@@ -363,3 +364,5 @@ require (
 )
 
 replace github.com/dapr/components-contrib => ../../
+
+replace github.com/dapr/dapr => github.com/sicoyle/dapr v0.0.0-20250721133219-ae723285c0e3
```
Comment (nelson-parente) [inline tests/certification/go.mod]: comment to make sure we don't forget to revert this

File: conversation/langchaingokit/model.go [hunk 5]
```diff
@@ -27,27 +27,56 @@ type LLM struct {
...
+
+	if r.Temperature > 0 {
+		opts = append(opts, llms.WithTemperature(r.Temperature))
+	}
+
+	if r.Tools != nil {
+		opts = append(opts, llms.WithTools(*r.Tools))
+	}
+
+	opts = append(opts, llms.WithToolChoice(r.ToolChoice))
```
Comment (nelson-parente) [inline conversation/langchaingokit/model.go]: what if `r.ToolChoice` is value is invalid?

File: conversation/deepseek/deepseek.go [hunk 6]
```diff
@@ -20,20 +20,26 @@ import (
 	"reflect"
 
 	"github.com/dapr/components-contrib/conversation"
+	"github.com/dapr/components-contrib/conversation/langchaingokit"
 	"github.com/dapr/components-contrib/metadata"
 	"github.com/dapr/kit/logger"
 	kmeta "github.com/dapr/kit/metadata"
 
-	deepseek_go "github.com/cohesion-org/deepseek-go"
+	"github.com/tmc/langchaingo/llms/openai"
```
Comment (nelson-parente) [inline conversation/deepseek/deepseek.go:28]: nit: organize imports

File: conversation/deepseek/deepseek.go [hunk 7]
```diff
@@ -60,50 +81,6 @@ func (d *Deepseek) GetComponentMetadata() (metadataInfo metadata.MetadataMap) {
 	return
 }
 
-func (d *Deepseek) Converse(ctx context.Context, r *conversation.ConversationRequest) (res *conversation.ConversationResponse, err error) {
```
Comment (nelson-parente) [inline conversation/deepseek/deepseek.go:63]: why are we removing this?

File: conversation/echo/echo.go [hunk 8]
```diff
@@ -22,6 +22,7 @@ import (
 	"github.com/dapr/components-contrib/metadata"
 	"github.com/dapr/kit/logger"
 	kmeta "github.com/dapr/kit/metadata"
+	"github.com/tmc/langchaingo/llms"
```
Comment (nelson-parente) [inline conversation/echo/echo.go]: nit: organize imports

File: conversation/echo/echo.go [hunk 9]
```diff
@@ -51,24 +53,43 @@ func (e *Echo) Init(ctx context.Context, meta conversation.Metadata) error {
...
+			switch p := part.(type) {
+			case llms.TextContent:
+				content += p.Text
+			case *llms.ToolCall:
+				if p.FunctionCall != nil {
+					content += p.FunctionCall.Name + "(" + p.FunctionCall.Arguments + ")"
+				}
+			case llms.ToolCallResponse:
+				content += p.Content
+			default:
```
Comment (nelson-parente) [inline conversation/echo/echo.go:86]: should this return an error?

**Review by nelson-parente — COMMENTED** [review]
overall looks good! please take a look at some of the TODO comments in the code

File: conversation/langchaingokit/model.go (hunk 5 above)
Comment (sicoyle) [reply conversation/langchaingokit/model.go]: dapr/dapr enforces a valid selection for this

File: conversation/deepseek/deepseek.go (hunk 6 above)
Comment (sicoyle) [reply conversation/deepseek/deepseek.go:28]: i ran the lint cmd for this :)

File: conversation/deepseek/deepseek.go (hunk 7 above)
Comment (sicoyle) [reply conversation/deepseek/deepseek.go:63]: This is removed by allowing deepseek to use the underlying langchain llms.LLM instead of having its own disjoint Converse logic. See the `langchaingokit.LLM` field on the Deepseek struct

File: conversation/echo/echo.go (hunk 9 above)
Comment (sicoyle) [reply conversation/echo/echo.go:86]: yes i can update to err here instead

**Comment by sicoyle** [comment]
> overall looks good! please take a look at some of the TODO comments in the code

done, thank you @nelson-parente ✅

File: conversation/langchaingokit/model.go (hunk 5 above)
Comment (sicoyle) [reply conversation/langchaingokit/model.go]: i went ahead and wrapped it in an if justttt in case, but yeah runtime enforces it is never empty :)

**Comment by sicoyle** [comment]
```
g) ✗ go test -v -tags=conftests -count=1 ./tests/conformance -run="TestConversationConformance/echo"
=== RUN   TestConversationConformance
=== RUN   TestConversationConformance/echo
=== RUN   TestConversationConformance/echo/init
=== RUN   TestConversationConformance/echo/converse
=== RUN   TestConversationConformance/echo/converse/get_a_non-empty_response_without_errors
=== RUN   TestConversationConformance/echo/converse/test_user_message_type
=== RUN   TestConversationConformance/echo/converse/test_system_message_type
=== RUN   TestConversationConformance/echo/converse/test_assistant_message_type
=== RUN   TestConversationConformance/echo/converse/test_developer_message_type
=== RUN   TestConversationConformance/echo/converse/test_tool_message_type_-_confirming_active_tool_calling_capability
=== RUN   TestConversationConformance/echo/converse/test_conversation_history_with_tool_calls
--- PASS: TestConversationConformance (0.00s)
    --- PASS: TestConversationConformance/echo (0.00s)
        --- PASS: TestConversationConformance/echo/init (0.00s)
        --- PASS: TestConversationConformance/echo/converse (0.00s)
            --- PASS: TestConversationConformance/echo/converse/get_a_non-empty_response_without_errors (0.00s)
            --- PASS: TestConversationConformance/echo/converse/test_user_message_type (0.00s)
            --- PASS: TestConversationConformance/echo/converse/test_system_message_type (0.00s)
            --- PASS: TestConversationConformance/echo/converse/test_assistant_message_type (0.00s)
            --- PASS: TestConversationConformance/echo/converse/test_developer_message_type (0.00s)
            --- PASS: TestConversationConformance/echo/converse/test_tool_message_type_-_confirming_active_tool_calling_capability (0.00s)
            --- PASS: TestConversationConformance/echo/converse/test_conversation_history_with_tool_calls (0.00s)
PASS
ok      github.com/dapr/components-contrib/tests/conformance    1.150s
➜  components-contrib git:(feat-conv-tool-calling) ✗ go test -v -tags=conftests -count=1 ./tests/conformance -run="TestConversationConformance/mistral"
=== RUN   TestConversationConformance
=== RUN   TestConversationConformance/mistral
=== RUN   TestConversationConformance/mistral/init
=== RUN   TestConversationConformance/mistral/converse
=== RUN   TestConversationConformance/mistral/converse/get_a_non-empty_response_without_errors
=== RUN   TestConversationConformance/mistral/converse/test_user_message_type
=== RUN   TestConversationConformance/mistral/converse/test_system_message_type
=== RUN   TestConversationConformance/mistral/converse/test_assistant_message_type
=== RUN   TestConversationConformance/mistral/converse/test_developer_message_type
=== RUN   TestConversationConformance/mistral/converse/test_tool_message_type_-_confirming_active_tool_calling_capability
=== RUN   TestConversationConformance/mistral/converse/test_conversation_history_with_tool_calls
--- PASS: TestConversationConformance (6.83s)
    --- PASS: TestConversationConformance/mistral (6.83s)
        --- PASS: TestConversationConformance/mistral/init (0.00s)
        --- PASS: TestConversationConformance/mistral/converse (6.83s)
            --- PASS: TestConversationConformance/mistral/converse/get_a_non-empty_response_without_errors (0.69s)
            --- PASS: TestConversationConformance/mistral/converse/test_user_message_type (0.72s)
            --- PASS: TestConversationConformance/mistral/converse/test_system_message_type (0.49s)
            --- PASS: TestConversationConformance/mistral/converse/test_assistant_message_type (1.46s)
            --- PASS: TestConversationConformance/mistral/converse/test_developer_message_type (1.16s)
            --- PASS: TestConversationConformance/mistral/converse/test_tool_message_type_-_confirming_active_tool_calling_capability (0.96s)
            --- PASS: TestConversationConformance/mistral/converse/test_conversation_history_with_tool_calls (1.34s)
PASS
ok      github.com/dapr/components-contrib/tests/conformance    7.977s
➜  components-contrib git:(feat-conv-tool-calling) ✗ go test -v -tags=conftests -count=1 ./tests/conformance -run="TestConversationConformance/openai" 
=== RUN   TestConversationConformance
=== RUN   TestConversationConformance/openai
=== RUN   TestConversationConformance/openai/init
=== RUN   TestConversationConformance/openai/converse
=== RUN   TestConversationConformance/openai/converse/get_a_non-empty_response_without_errors
=== RUN   TestConversationConformance/openai/converse/test_user_message_type
=== RUN   TestConversationConformance/openai/converse/test_system_message_type
=== RUN   TestConversationConformance/openai/converse/test_assistant_message_type
=== RUN   TestConversationConformance/openai/converse/test_developer_message_type
=== RUN   TestConversationConformance/openai/converse/test_tool_message_type_-_confirming_active_tool_calling_capability
=== RUN   TestConversationConformance/openai/converse/test_conversation_history_with_tool_calls
--- PASS: TestConversationConformance (10.72s)
    --- PASS: TestConversationConformance/openai (10.72s)
        --- PASS: TestConversationConformance/openai/init (0.00s)
        --- PASS: TestConversationConformance/openai/converse (10.72s)
            --- PASS: TestConversationConformance/openai/converse/get_a_non-empty_response_without_errors (1.93s)
            --- PASS: TestConversationConformance/openai/converse/test_user_message_type (1.04s)
            --- PASS: TestConversationConformance/openai/converse/test_system_message_type (0.63s)
            --- PASS: TestConversationConformance/openai/converse/test_assistant_message_type (0.82s)
            --- PASS: TestConversationConformance/openai/converse/test_developer_message_type (2.04s)
            --- PASS: TestConversationConformance/openai/converse/test_tool_message_type_-_confirming_active_tool_calling_capability (1.54s)
            --- PASS: TestConversationConformance/openai/converse/test_conversation_history_with_tool_calls (2.72s)
PASS
ok      github.com/dapr/components-contrib/tests/conformance    11.975s
➜  components-contrib git:(feat-conv-tool-calling) ✗ go test -v -tags=conftests -count=1 ./tests/conformance -run="TestConversationConformance/anthropic"
=== RUN   TestConversationConformance
=== RUN   TestConversationConformance/anthropic
=== RUN   TestConversationConformance/anthropic/init
=== RUN   TestConversationConformance/anthropic/converse
=== RUN   TestConversationConformance/anthropic/converse/get_a_non-empty_response_without_errors
=== RUN   TestConversationConformance/anthropic/converse/test_user_message_type
=== RUN   TestConversationConformance/anthropic/converse/test_system_message_type
=== RUN   TestConversationConformance/anthropic/converse/test_assistant_message_type
=== RUN   TestConversationConformance/anthropic/converse/test_developer_message_type
=== RUN   TestConversationConformance/anthropic/converse/test_tool_message_type_-_confirming_active_tool_calling_capability
=== RUN   TestConversationConformance/anthropic/converse/test_conversation_history_with_tool_calls
--- PASS: TestConversationConformance (6.90s)
    --- PASS: TestConversationConformance/anthropic (6.90s)
        --- PASS: TestConversationConformance/anthropic/init (0.00s)
        --- PASS: TestConversationConformance/anthropic/converse (6.90s)
            --- PASS: TestConversationConformance/anthropic/converse/get_a_non-empty_response_without_errors (1.84s)
            --- PASS: TestConversationConformance/anthropic/converse/test_user_message_type (0.76s)
            --- PASS: TestConversationConformance/anthropic/converse/test_system_message_type (0.41s)
            --- PASS: TestConversationConformance/anthropic/converse/test_assistant_message_type (0.68s)
            --- PASS: TestConversationConformance/anthropic/converse/test_developer_message_type (0.96s)
            --- PASS: TestConversationConformance/anthropic/converse/test_tool_message_type_-_confirming_active_tool_calling_capability (0.72s)
            --- PASS: TestConversationConformance/anthropic/converse/test_conversation_history_with_tool_calls (1.53s)
PASS
ok      github.com/dapr/components-contrib/tests/conformance    10.759s
```

**Review by nelson-parente — APPROVED** [review]

**Comment by sicoyle** [comment]
unrelated failures

```
github.com/dapr/components-contrib/bindings/zeebe	0.009s	coverage: 43.8% of statements
--- FAIL: TestActivateJobs (0.00s)
    --- FAIL: TestActivateJobs/maxJobsToActivate_is_mandatory (0.00s)
        activate_jobs_test.go:124: 
            	Error Trace:	/home/runner/work/components-contrib/components-contrib/bindings/zeebe/command/activate_jobs_test.go:124
            	Error:      	Target error should be in err chain:
            	            	expected: "maxJobsToActivate is a required attribute"
            	            	in chain: "invalid duration"
            	Test:       	TestActivateJobs/maxJobsToActivate_is_mandatory
    --- FAIL: TestActivateJobs/activate_jobs_with_mandatory_fields (0.00s)
        activate_jobs_test.go:141: 
            	Error Trace:	/home/runner/work/components-contrib/components-contrib/bindings/zeebe/command/activate_jobs_test.go:141
            	Error:      	Received unexpected error:
            	            	invalid duration
            	Test:       	TestActivateJobs/activate_jobs_with_mandatory_fields
    --- FAIL: TestActivateJobs/send_message_with_optional_fields (0.00s)
        activate_jobs_test.go:1[64](https://github.com/dapr/components-contrib/actions/runs/16444948550/job/46474332883?pr=3890#step:9:65): 
            	Error Trace:	/home/runner/work/components-contrib/components-contrib/bindings/zeebe/command/activate_jobs_test.go:164
            	Error:      	Received unexpected error:
            	            	invalid duration
            	Test:       	TestActivateJobs/send_message_with_optional_fields
--- FAIL: TestCreateInstance (0.00s)
    --- FAIL: TestCreateInstance/bpmnProcessId_and_processDefinitionKey_are_not_allowed_at_the_same_time (0.00s)
        create_instance_test.go:118: 
            	Error Trace:	/home/runner/work/components-contrib/components-contrib/bindings/zeebe/command/create_instance_test.go:118
            	Error:      	Target error should be in err chain:
            	            	expected: "either 'bpmnProcessId' or 'processDefinitionKey' must be passed, not both at the same time"
            	            	in chain: "invalid duration"
            	Test:       	TestCreateInstance/bpmnProcessId_and_processDefinitionKey_are_not_allowed_at_the_same_time
    --- FAIL: TestCreateInstance/either_bpmnProcessId_or_processDefinitionKey_must_be_given (0.00s)
        create_instance_test.go:132: 
            	Error Trace:	/home/runner/work/components-contrib/components-contrib/bindings/zeebe/command/create_instance_test.go:132
            	Error:      	Target error should be in err chain:
            	            	expected: "either 'bpmnProcessId' or 'processDefinitionKey' must be passed"
            	            	in chain: "invalid duration"
            	Test:       	TestCreateInstance/either_bpmnProcessId_or_processDefinitionKey_must_be_given
    --- FAIL: TestCreateInstance/create_command_with_bpmnProcessId_and_specific_version (0.00s)
        create_instance_test.go:149: 
            	Error Trace:	/home/runner/work/components-contrib/components-contrib/bindings/zeebe/command/create_instance_test.go:149
            	Error:      	Received unexpected error:
            	            	invalid duration
            	Test:       	TestCreateInstance/create_command_with_bpmnProcessId_and_specific_version
    --- FAIL: TestCreateInstance/create_command_with_bpmnProcessId_and_latest_version (0.00s)
        create_instance_test.go:1[68](https://github.com/dapr/components-contrib/actions/runs/16444948550/job/46474332883?pr=3890#step:9:69): 
            	Error Trace:	/home/runner/work/components-contrib/components-contrib/bindings/zeebe/command/create_instance_test.go:168
            	Error:      	Received unexpected error:
            	            	invalid duration
            	Test:       	TestCreateInstance/create_command_with_bpmnProcessId_and_latest_version
    --- FAIL: TestCreateInstance/create_command_with_processDefinitionKey (0.00s)
        create_instance_test.go:187: 
            	Error Trace:	/home/runner/work/components-contrib/components-contrib/bindings/zeebe/command/create_instance_test.go:187
            	Error:      	Received unexpected error:
            	            	invalid duration
            	Test:       	TestCreateInstance/create_command_with_processDefinitionKey
    --- FAIL: TestCreateInstance/create_command_with_variables (0.00s)
        create_instance_test.go:208: 
            	Error Trace:	/home/runner/work/components-contrib/components-contrib/bindings/zeebe/command/create_instance_test.go:208
            	Error:      	Received unexpected error:
            	            	invalid duration
            	Test:       	TestCreateInstance/create_command_with_variables
--- FAIL: TestFailJob (0.00s)
    --- FAIL: TestFailJob/retries_is_mandatory (0.00s)
        fail_job_test.go:108: 
            	Error Trace:	/home/runner/work/components-contrib/components-contrib/bindings/zeebe/command/fail_job_test.go:108
            	Error:      	Target error should be in err chain:
            	            	expected: "retries is a required attribute"
            	            	in chain: "invalid duration"
            	Test:       	TestFailJob/retries_is_mandatory
    --- FAIL: TestFailJob/fail_a_job (0.00s)
        fail_job_test.go:126: 
            	Error Trace:	/home/runner/work/components-contrib/components-contrib/bindings/zeebe/command/fail_job_test.go:126
            	Error:      	Received unexpected error:
            	            	invalid duration
            	Test:       	TestFailJob/fail_a_job
--- FAIL: TestPublishMessage (0.00s)
    --- FAIL: TestPublishMessage/send_message_with_mandatory_fields (0.00s)
        publish_message_test.go:128: 
            	Error Trace:	/home/runner/work/components-contrib/components-contrib/bindings/zeebe/command/publish_message_test.go:128
            	Error:      	Received unexpected error:
            	            	invalid duration
            	Test:       	TestPublishMessage/send_message_with_mandatory_fields
    --- FAIL: TestPublishMessage/send_message_with_optional_fields (0.00s)
        publish_message_test.go:153: 
            	Error Trace:	/home/runner/work/components-contrib/components-contrib/bindings/zeebe/command/publish_message_test.go:153
            	Error:      	Received unexpected error:
            	            	invalid duration
            	Test:       	TestPublishMessage/send_message_with_optional_fields
FAIL
```

**Comment by sicoyle** [comment]
Everything in the build is passing in this last commit except:

The kafka cert test fails in this PR, but seems like it has with the same err in other PRs too so not related to my changes at all.
https://github.com/dapr/components-contrib/actions/runs/16193984270/job/45740849890
https://github.com/dapr/components-contrib/pull/

even here it fails
https://github.com/dapr/components-contrib/actions/runs/16445867297/job/46477573796
https://github.com/dapr/components-contrib/pull/3897

**Review by nelson-parente — APPROVED** [review]

**Comment by sicoyle** [comment]
the plan is merge here and i will come back to fix the circular dependency in the cert test go mod after i get the dapr/dapr pr merged for this

**Review by yaron2 — APPROVED** [review]

**Comment by msfussell** [comment]
Docs https://github.com/dapr/docs/issues/4687
